### PR TITLE
Adapting the LaTeX renderer to use the new AST

### PR DESCRIPTION
**Pull request by GarkGarcia**
A follow up to #19. Started adapting the `LatexBuilder` class to the new AST.

One issue we're facing is that the new AST does not distinguish symbols (such as `x`) from sections of text (such as `"x"`) - they are both represented as strings. This distinction is pretty relevant for LaTeX, so we might need to rethink that bit.

This could be fixed  by parsing sections of test as 

```ruby
{:type => :text, :e => "some text"}
```
instead of simply

```ruby
"some text"
```

Despite this, the new AST works great. The current LaTeX renderer passes all test that doens't involve text.

### CODE REVIEW

**Comment by pepijnve** [comment]
I ran into the same text issue with MathML. For now I used length > 1 as the criterium to choose between `mi` and `mtext`. Not 100% accurate of course.

**Comment by GarkGarcia** [comment]
I think all entries in the [Symbols Table](https://github.com/asciidoctor/asciimath/blob/ast/AST.adoc#symbol-table) is now covered.

**Comment by GarkGarcia** [comment]
> I ran into the same text issue with MathML. For now I used length > 1 as the criterium to choose between `mi` and `mtext`. Not 100% accurate of course since you can't have single character text strings this way. 

I see. Is there a particular reason for not tagging text-blocks? I could work on this change.

**Comment by pepijnve** [comment]
> Is there a particular reason for not tagging text-blocks?

Not really. I design/code in an exploratory way. This is just the first revision I came up with.

It might be a good idea to extend this tagging to other AST node types as well. In the MathML backend for instance, I've had to match text against a number regex to know if I should output `mn` or `mi`. The parser already does the exact same thing, but discards that information.

It would be nice to make the whole thing consistent then (i.e. Hashes with a type and Arrays for everything), but that might cumbersome to use for the Symbols.

File: lib/asciimath/latex.rb [hunk 1]
```diff
@@ -107,7 +107,13 @@ def append(expression, separator = " ")
             @latex << separator if i != len
           end
         when String
-          append_escaped(expression)
+          if expression.length > 1 and not expression.match(/^[1-9](\.[1-9]+)?/)
```
Comment (pepijnve) [inline lib/asciimath/latex.rb:112]: I think you want `[0-9]` instead of `[1-9]` here.

**Comment by GarkGarcia** [comment]
> It might be a good idea to extend this tagging to other AST node types as well. In the MathML backend for instance, I've had to match text against a number regex to know if I should output `mn` or `mi`. The parser already does the exact same thing, but discards that information.
> 
> It would be nice to make the whole thing consistent then (i.e. Hashes with a type and Arrays for everything), but that might cumbersome to use for the Symbols.

I agree. Tagging everything (besides the arrays) is probably a good idea.

In regards to the separation between identifiers and operators, I wouldn't mind the separation if they were to be represented in the exact same way (with only the `:type`attribute changing), so that I could match on `:identifier, :operator` and treat them as the same.

We could also provide some hinting to help you interpreting the meaning of each symbol, but I prefer the idea mentioned above.

**Comment by GarkGarcia** [comment]
Would the following representations work for MathML?

```ruby
{:type => :identifier, :e => "<String or Symbol>"}
```

```ruby
{:type => :operator, :e => "<String or Symbol>"}
```

**Comment by pepijnve** [comment]
I ended up with something close to what you proposed.

```
{:type => :symbol, :value => Symbol}
{:type => :identifier, :value => String}
{:type => :text, :value => String}
{:type => :number, :value => String}
```

`:text` and `:number` are self explanatory I think.
`:symbol` is used for any recognised symbol in the symbol table.
`:identifier` is used for whatever's left.

The mapping for MathML is
```
:text => <mtext>
:number => <mn>
:identifier => <mi>
:symbol => depends on what the symbol represents
```

Note that I did not wrap the parentheses symbols for `:paren` and `:matrix` in a symbol node. Those are just the unwrapped Symbols.

**Comment by pepijnve** [comment]
I think I'm done with the AST work at this point, the AST, MathML and HTML specs are all passing now and I'm pretty happy with the end result. I promise I'll stop moving your target now 😄.

**Comment by GarkGarcia** [comment]
> I think I'm done with the AST work at this point. The AST, MathML and HTML specs are all passing now and I'm pretty happy with the end result. I promise I'll stop moving your target now smile.

Great! I'm happy with the results too, and the LaTeX spec is passing every test 😁️

**Comment by GarkGarcia** [comment]
I'm finished with the commit.

**Pull request merged**